### PR TITLE
[Feature][RayCluster]: introduce RayClusterSuspending and RayClusterSuspended conditions

### PR DESCRIPTION
**Pull request by rueian**
## Why are these changes needed?

Resolves https://github.com/ray-project/kuberay/issues/2277

~~This PR introduces a new status condition: `RayClusterSuspend` with two variations, `RayClusterSuspending` and `RayClusterSuspended`.~~

This PR introduces two new status conditions: `RayClusterSuspending` and `RayClusterSuspended`.

The new conditions will populated by default with `metav1.ConditionFalse` if the `features.Enabled(features.RayClusterStatusConditions)` is enabled.

When the gate is enabled, we rely on the new `RayClusterSuspending` to suspend a RayCluster atomically. When a user turns on the `.Spec.Suspend` of a RayCluster, we will enable the `RayClusterSuspending` and persist it before actually stopping all Pods. We will stop all Pods in the next reconciliation and do not consider the `.Spec.Suspend` again before switching to the `RayClusterSuspend`.

If the condition `RayClusterSuspend` is enabled and the `.Spec.Suspend` is turned off by the user, we will disable the condition by setting it to `metav1.ConditionFalse`, so that we can preserve the transition timestamp for more observability.

In summary, this is the status BEFORE suspending a RayCluster:

image

This is the status DURING suspending a RayCluster:

image

This is the status AFTER it is suspended:

image

This is the status AFTER the cluster is resumed:

image

## Related issue number
https://github.com/ray-project/kuberay/issues/2277

## Checks
- [x] I've made sure the tests are passing.
* Testing Strategy
  - [x] Unit tests
  - [x] Manual tests

### CODE REVIEW

**Comment by rueian** [comment]
Hi @kevin85421, please take a look when you are available.

**Comment by kevin85421** [comment]
I will review the PR tomorrow.

File: ray-operator/controllers/ray/raycluster_controller.go [hunk 1]
```diff
@@ -1170,15 +1190,27 @@ func (r *RayClusterReconciler) SetupWithManager(mgr ctrl.Manager, reconcileConcu
...
+	if reconcileErr != nil && !errstd.Is(reconcileErr, errSuspendCondition) {
 		logger.Info("Reconciliation error", "error", reconcileErr)
 	}
 
 	// Deep copy the instance, so we don't mutate the original object.
 	newInstance := instance.DeepCopy()
 
 	if features.Enabled(features.RayClusterStatusConditions) {
-		if reconcileErr != nil {
+		if errstd.Is(reconcileErr, errSuspending) {
```
Comment (kevin85421) [inline ray-operator/controllers/ray/raycluster_controller.go:1201]: What’s the value of `Reason` for the RayClusterSuspend condition if we set `spec.Suspend` to true and then back to false?

File: ray-operator/controllers/ray/raycluster_controller.go [hunk 2]
```diff
@@ -614,8 +619,10 @@ func (r *RayClusterReconciler) reconcileHeadlessService(ctx context.Context, ins
 func (r *RayClusterReconciler) reconcilePods(ctx context.Context, instance *rayv1.RayCluster) error {
 	logger := ctrl.LoggerFrom(ctx)
 
-	// if RayCluster is suspended, delete all pods and skip reconcile
-	if instance.Spec.Suspend != nil && *instance.Spec.Suspend {
+	// if RayCluster is suspending, delete all pods and skip reconcile
+	suspendStatus := utils.FindRayClusterSuspendStatus(instance)
+	if suspendStatus == rayv1.RayClusterSuspending ||
+		(!features.Enabled(features.RayClusterStatusConditions) && instance.Spec.Suspend != nil && *instance.Spec.Suspend) {
 		if _, err := r.deleteAllPods(ctx, common.RayClusterAllPodsAssociationOptions(instance)); err != nil {
```
Comment (kevin85421) [inline ray-operator/controllers/ray/raycluster_controller.go:626]: Do we need to send a `deleteAllPods` request to the K8s API server for each reconciliation during `Suspending`?

File: ray-operator/controllers/ray/raycluster_controller.go [hunk 3]
```diff
@@ -614,8 +619,10 @@ func (r *RayClusterReconciler) reconcileHeadlessService(ctx context.Context, ins
 func (r *RayClusterReconciler) reconcilePods(ctx context.Context, instance *rayv1.RayCluster) error {
 	logger := ctrl.LoggerFrom(ctx)
 
-	// if RayCluster is suspended, delete all pods and skip reconcile
-	if instance.Spec.Suspend != nil && *instance.Spec.Suspend {
+	// if RayCluster is suspending, delete all pods and skip reconcile
+	suspendStatus := utils.FindRayClusterSuspendStatus(instance)
+	if suspendStatus == rayv1.RayClusterSuspending ||
+		(!features.Enabled(features.RayClusterStatusConditions) && instance.Spec.Suspend != nil && *instance.Spec.Suspend) {
```
Comment (kevin85421) [inline ray-operator/controllers/ray/raycluster_controller.go:625]: Both L625 and L639 use `features.Enabled(features.RayClusterStatusConditions)`. Can we use a variable instead?

File: ray-operator/controllers/ray/raycluster_controller.go [hunk 4]
```diff
@@ -629,6 +636,19 @@ func (r *RayClusterReconciler) reconcilePods(ctx context.Context, instance *rayv
 		return nil
 	}
 
+	if features.Enabled(features.RayClusterStatusConditions) {
+		if suspendStatus == rayv1.RayClusterSuspended {
+			if instance.Spec.Suspend != nil && !*instance.Spec.Suspend {
```
Comment (kevin85421) [inline ray-operator/controllers/ray/raycluster_controller.go:641]: Do we need to return errors (`errResuming` or `errSuspending`) here? Can we return `nil` directly? We can directly infer the cluster is suspending, or suspended, or resuming in `calculateStatus`. Removing the errors can also avoid handling special cases in `calculateStatus`.

File: ray-operator/controllers/ray/raycluster_controller.go [hunk 5]
```diff
@@ -55,6 +55,11 @@ var (
 
 	// Definition of a index field for pod name
 	podUIDIndexField = "metadata.uid"
+
```
Comment (kevin85421) [inline ray-operator/controllers/ray/raycluster_controller.go:58]: remove these errors

File: ray-operator/controllers/ray/raycluster_controller.go [hunk 6]
```diff
@@ -1170,15 +1190,27 @@ func (r *RayClusterReconciler) SetupWithManager(mgr ctrl.Manager, reconcileConcu
 func (r *RayClusterReconciler) calculateStatus(ctx context.Context, instance *rayv1.RayCluster, reconcileErr error) (*rayv1.RayCluster, error) {
```
Comment (kevin85421) [inline ray-operator/controllers/ray/raycluster_controller.go:1190]: Could we avoid handing the special case for `errSuspendCondition`? Instead, we can infer the values of suspending, suspended, or resuming.

* `suspending`
  * false by default
  * false -> true: `suspended` is false and `spec.suspend` is true, and clean up related fields in `status`.
  * true -> false: all Pods are deleted, and set `suspended` from false to true.
* `suspended`
  *  false by default
  * false -> true: when `suspending` transitions from true to false
  * true -> false: `spec.Suspend` is false

**Comment by kevin85421** [comment]
Btw, we may also ping some Kueue folks to review this PR before merge.

**Review by kevin85421 — COMMENTED** [review]
Discussion offline:

1. Use two conditions: `suspending` and `suspended`.
2. Return a boolean to determine whether to requeue the CR immediately.

File: ray-operator/controllers/ray/raycluster_controller.go [hunk 7]
```diff
@@ -1514,17 +1572,18 @@ func (r *RayClusterReconciler) reconcileAutoscalerRoleBinding(ctx context.Contex
 	return nil
 }
 
-func (r *RayClusterReconciler) updateRayClusterStatus(ctx context.Context, originalRayClusterInstance, newInstance *rayv1.RayCluster) error {
+func (r *RayClusterReconciler) updateRayClusterStatus(ctx context.Context, originalRayClusterInstance, newInstance *rayv1.RayCluster) (bool, error) {
```
Comment (kevin85421) [inline ray-operator/controllers/ray/raycluster_controller.go:1575]: can you add a comment to describe the return values of this function?

File: ray-operator/controllers/ray/raycluster_controller.go [hunk 8]
```diff
@@ -355,7 +356,7 @@ func (r *RayClusterReconciler) rayClusterReconcile(ctx context.Context, request
 	} else {
 		err = updateErr
 	}
-	if err != nil {
+	if err != nil || inconsistent {
```
Comment (kevin85421) [inline ray-operator/controllers/ray/raycluster_controller.go:359]: can you add a comment to explain why do we need to requeue the CR if the CR status is updated?

File: ray-operator/controllers/ray/raycluster_controller.go [hunk 9]
```diff
@@ -340,10 +340,11 @@ func (r *RayClusterReconciler) rayClusterReconcile(ctx context.Context, request
 	// Calculate the new status for the RayCluster. Note that the function will deep copy `instance` instead of mutating it.
```
Comment (kevin85421) [inline ray-operator/controllers/ray/raycluster_controller.go:340]: Can you validate the status of the RayCluster CR, `validateRayClusterStatus`, at the very beginning of the reconciliation to ensure that `RayClusterSuspending` and `RayClusterSuspended` are not both true at the same time? You can refer to [validateRayJobStatus](https://github.com/ray-project/kuberay/blob/bee1b71e01785aaa29c6ec9a0ad6604eb868579c/ray-operator/controllers/ray/rayjob_controller.go#L841) as an example.

File: ray-operator/controllers/ray/raycluster_controller_test.go [hunk 10]
```diff
@@ -479,6 +486,16 @@ var _ = Context("Inside the default namespace", func() {
...
 				time.Second*3, time.Millisecond*500).Should(Equal(rayv1.Suspended))
+			if withConditionEnabled {
+				Eventually(
+					func() rayv1.RayClusterConditionType {
+						if err := getResourceFunc(ctx, client.ObjectKey{Name: rayCluster.Name, Namespace: namespace}, rayCluster)(); err != nil {
+							return ""
+						}
+						return utils.FindRayClusterSuspendStatus(rayCluster)
+					},
+					time.Second*10, time.Millisecond*1000).Should(Equal(rayv1.RayClusterSuspended))
```
Comment (kevin85421) [inline ray-operator/controllers/ray/raycluster_controller_test.go:497]: nit: just make the parameters consistent with others
```suggestion
					time.Second*3, time.Millisecond*500).Should(Equal(rayv1.RayClusterSuspended))
```

File: ray-operator/controllers/ray/raycluster_controller_test.go [hunk 11]
```diff
@@ -534,6 +551,17 @@ var _ = Context("Inside the default namespace", func() {
...
+
+			if withConditionEnabled {
+				Eventually(
+					func() rayv1.RayClusterConditionType {
+						if err := getResourceFunc(ctx, client.ObjectKey{Name: rayCluster.Name, Namespace: namespace}, rayCluster)(); err != nil {
+							return ""
+						}
+						return utils.FindRayClusterSuspendStatus(rayCluster)
+					},
+					time.Second*10, time.Millisecond*1000).Should(Equal(rayv1.RayClusterSuspended))
```
Comment (kevin85421) [inline ray-operator/controllers/ray/raycluster_controller_test.go:563]: nit: just make the parameters consistent with others
```suggestion
					time.Second*3, time.Millisecond*500).Should(Equal(rayv1.RayClusterSuspended))
```

File: ray-operator/controllers/ray/raycluster_controller_test.go [hunk 12]
```diff
@@ -479,6 +486,16 @@ var _ = Context("Inside the default namespace", func() {
 			Eventually(
 				getClusterState(ctx, namespace, rayCluster.Name),
 				time.Second*3, time.Millisecond*500).Should(Equal(rayv1.Suspended))
+			if withConditionEnabled {
+				Eventually(
+					func() rayv1.RayClusterConditionType {
```
Comment (kevin85421) [inline ray-operator/controllers/ray/raycluster_controller_test.go:491]: This anonymous function is defined three times in this PR. Could we define it once and reuse it?

File: ray-operator/controllers/ray/raycluster_controller_test.go [hunk 13]
```diff
@@ -534,6 +551,17 @@ var _ = Context("Inside the default namespace", func() {
 			Eventually(
 				getClusterState(ctx, namespace, rayCluster.Name),
 				time.Second*3, time.Millisecond*500).Should(Equal(rayv1.Suspended))
+
```
Comment (kevin85421) [inline ray-operator/controllers/ray/raycluster_controller_test.go:554]: Could you check that both `suspending` and `suspended` should be false when `suspend` transitions from true to false?

File: ray-operator/controllers/ray/raycluster_controller.go [hunk 14]
```diff
@@ -1257,6 +1272,49 @@ func (r *RayClusterReconciler) calculateStatus(ctx context.Context, instance *ra
 			}
 		}
 
+		suspendStatus := utils.FindRayClusterSuspendStatus(newInstance)
+		if suspendStatus == rayv1.RayClusterSuspending {
+			if len(runtimePods.Items) == 0 {
+				meta.RemoveStatusCondition(&newInstance.Status.Conditions, string(rayv1.RayClusterProvisioned))
```
Comment (kevin85421) [inline ray-operator/controllers/ray/raycluster_controller.go:1278]: how about setting it back to false?

File: ray-operator/controllers/ray/raycluster_controller.go [hunk 15]
```diff
@@ -1257,6 +1272,49 @@ func (r *RayClusterReconciler) calculateStatus(ctx context.Context, instance *ra
...
+				meta.SetStatusCondition(&newInstance.Status.Conditions, metav1.Condition{
+					Type:   string(rayv1.RayClusterSuspending),
+					Reason: string(rayv1.RayClusterSuspending),
+					Status: metav1.ConditionFalse,
+				})
+				meta.SetStatusCondition(&newInstance.Status.Conditions, metav1.Condition{
```
Comment (kevin85421) [inline ray-operator/controllers/ray/raycluster_controller.go:1284]: We may need to consider which information in the CR status should be cleaned up. For example, https://github.com/ray-project/kuberay/blob/bee1b71e01785aaa29c6ec9a0ad6604eb868579c/ray-operator/controllers/ray/rayjob_controller.go#L314-L322

File: ray-operator/controllers/ray/raycluster_controller.go [hunk 16]
```diff
@@ -1257,6 +1272,49 @@ func (r *RayClusterReconciler) calculateStatus(ctx context.Context, instance *ra
 			}
 		}
 
+		suspendStatus := utils.FindRayClusterSuspendStatus(newInstance)
```
Comment (kevin85421) [inline ray-operator/controllers/ray/raycluster_controller.go:1275]: Can you add comments to explain the logic behind the status transitions here for future reference? Something similar to: https://github.com/ray-project/kuberay/pull/2403#discussion_r1788352417

**Review by kevin85421 — COMMENTED** [review]
The main motivation to add `suspending` is to make the `suspend` operation to be atomic. Can you add a test to test the atomicity of the `suspend` operation? You can refer to https://github.com/ray-project/kuberay/blob/bee1b71e01785aaa29c6ec9a0ad6604eb868579c/ray-operator/controllers/ray/rayjob_controller_suspended_test.go#L105

File: ray-operator/controllers/ray/raycluster_controller.go (hunk 2 above)
Comment (rueian) [reply ray-operator/controllers/ray/raycluster_controller.go:639]: Yes, we do that until all pods are deleted and then we switch to `suspended`.

File: ray-operator/controllers/ray/raycluster_controller_test.go (hunk 13 above)
Comment (rueian) [reply ray-operator/controllers/ray/raycluster_controller_test.go:619]: This is checked by the new `findRayClusterSuspendStatus` helper in tests.

File: ray-operator/controllers/ray/raycluster_controller.go (hunk 15 above)
Comment (rueian) [reply ray-operator/controllers/ray/raycluster_controller.go:1305]: I think the only fields we left are `HeadInfo.PodIP` and `HeadInfo.PodName`. We can update them in the `func (r *RayClusterReconciler) updateHeadInfo`:

```go
func (r *RayClusterReconciler) updateHeadInfo(ctx context.Context, instance *rayv1.RayCluster) error {
	headPod, err := common.GetRayClusterHeadPod(ctx, r, instance)
	if err != nil {
		return err
	}
	if headPod != nil {
		instance.Status.Head.PodIP = headPod.Status.PodIP
		instance.Status.Head.PodName = headPod.Name
	} else {
		instance.Status.Head.PodIP = ""   // <--- add this
		instance.Status.Head.PodName = "" // <--- add this
	}
```

I also added a test for checking whether the `Status.Head.PodName` is cleared. And `Status.Head.PodIP` is not tested because we don't have IPs in envtest.

File: ray-operator/controllers/ray/raycluster_controller.go [hunk 17]
```diff
@@ -1359,6 +1438,9 @@ func (r *RayClusterReconciler) updateHeadInfo(ctx context.Context, instance *ray
 	if headPod != nil {
 		instance.Status.Head.PodIP = headPod.Status.PodIP
 		instance.Status.Head.PodName = headPod.Name
+	} else {
+		instance.Status.Head.PodIP = ""
+		instance.Status.Head.PodName = ""
```
Comment (rueian) [inline ray-operator/controllers/ray/raycluster_controller.go:1443]: Clean the Head Pod information from the `Status.Head` when there is no Head Pod, ex: cluster suspended.

**Comment by rueian** [comment]
> The main motivation to add `suspending` is to make the `suspend` operation to be atomic. Can you add a test to test the atomicity of the `suspend` operation? You can refer to
> 
> https://github.com/ray-project/kuberay/blob/bee1b71e01785aaa29c6ec9a0ad6604eb868579c/ray-operator/controllers/ray/rayjob_controller_suspended_test.go#L105

Ok, I added it with the same finalizer technique to the existing test. And all the above suggestions are resolved too.

File: ray-operator/controllers/ray/utils/util.go [hunk 18]
```diff
@@ -102,6 +102,33 @@ func FindHeadPodReadyCondition(headPod *corev1.Pod) metav1.Condition {
...
+// The two conditions should not be both True at the same time. The transition logic should be the following:
+//
+//	rayv1.RayClusterSuspending:
+//	  False by default
+//	  False -> True: when Spec.Suspend is true.
+//	  True -> False: when all Pods are deleted, and also set rayv1.RayClusterSuspended from False to True.
+//	rayv1.RayClusterSuspended
+//	  False by default
+//	  False -> True: when suspending transitions from True to False
+//	  True -> False: when spec.Suspend is false.
```
Comment (kevin85421) [inline ray-operator/controllers/ray/utils/util.go:118]: nit
```suggestion
//	  True -> False: when `Spec.Suspend` is false.
```

File: ray-operator/controllers/ray/utils/util.go [hunk 19]
```diff
@@ -102,6 +102,33 @@ func FindHeadPodReadyCondition(headPod *corev1.Pod) metav1.Condition {
...
 
+// FindRayClusterSuspendStatus returns the current suspend status from two conditions:
+//  1. rayv1.RayClusterSuspending
+//  2. rayv1.RayClusterSuspended
+//
+// The two conditions should not be both True at the same time. The transition logic should be the following:
+//
+//	rayv1.RayClusterSuspending:
+//	  False by default
+//	  False -> True: when Spec.Suspend is true.
```
Comment (kevin85421) [inline ray-operator/controllers/ray/utils/util.go:113]: nit
```suggestion
//	  False -> True: when `Spec.Suspend` is true.
```

File: ray-operator/controllers/ray/raycluster_controller_test.go [hunk 20]
```diff
@@ -463,6 +498,46 @@ var _ = Context("Inside the default namespace", func() {
 			})
 			Expect(err).NotTo(HaveOccurred(), "Failed to update RayCluster")
 
+			// Make sure the suspending process is atomic by removing finalizers one by one.
+			if withConditionEnabled {
+				// Make sure the cluster is now marked with RayClusterSuspending.
+				Eventually(findRayClusterSuspendStatus, time.Second*3, time.Millisecond*500).Should(Equal(rayv1.RayClusterSuspending))
+				// Then turn the Spec.Suspend to false.
+				err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
```
Comment (kevin85421) [inline ray-operator/controllers/ray/raycluster_controller_test.go:506]: Can you create a function for updating `spec.Suspend` in case we call it twice? Maybe you can refer to `updateRayJobSuspendField`.

File: ray-operator/controllers/ray/raycluster_controller_test.go [hunk 21]
```diff
@@ -463,6 +498,46 @@ var _ = Context("Inside the default namespace", func() {
...
+					Eventually(
+						getResourceFunc(ctx, client.ObjectKey{Name: rayCluster.Name, Namespace: namespace}, rayCluster),
+						time.Second*3, time.Millisecond*500).Should(BeNil(), "rayCluster: %v", rayCluster)
+					suspend := false
+					rayCluster.Spec.Suspend = &suspend
+					return k8sClient.Update(ctx, rayCluster)
+				})
+				Expect(err).NotTo(HaveOccurred(), "Failed to update RayCluster")
+
+				// Remove the finalizers one by one to make sure the suspending process is still ongoing.
```
Comment (kevin85421) [inline ray-operator/controllers/ray/raycluster_controller_test.go:516]: Can we separate the atomicity test into its own test? I want to avoid mixing both code paths in a single test. In the atomicity test, can you verify the following:

1. Add finalizers to the Pods.
2. Set `suspend` to true, and check that the `suspending` condition is true.
3. Set `suspend` back to false, and ensure the `suspending` condition remains true for 2 to 3 seconds (use `Consistently`).
4. Remove the finalizers, and verify that both `suspending` and `suspended` conditions become false, all old Pods are deleted, and new Pods are created.
5. Set `suspend` to true, and all Pods should be deleted.

File: ray-operator/controllers/ray/raycluster_controller.go (hunk 2 above)
Comment (rueian) [reply ray-operator/controllers/ray/raycluster_controller.go:640]: After revising the source, I found that `deleteAllPods` already did the improvement: only sends API requests if there are Pods still running.
https://github.com/ray-project/kuberay/blob/35e913a9b96e84b8e4a41ebe1efc6e656420afbd/ray-operator/controllers/ray/raycluster_controller.go#L195-L210

I guess there is no need for an issue to track the progress?

File: ray-operator/controllers/ray/raycluster_controller_test.go (hunk 21 above)
Comment (rueian) [reply ray-operator/controllers/ray/raycluster_controller_test.go]: ok, a new test is added.

**Review by kevin85421 — APPROVED** [review]

File: ray-operator/controllers/ray/suite_helpers_test.go [hunk 22]
```diff
@@ -33,7 +33,7 @@ func listResourceFunc(ctx context.Context, workerPods *corev1.PodList, opt ...cl
 
 		count := 0
 		for _, aPod := range workerPods.Items {
-			if (reflect.DeepEqual(aPod.Status.Phase, corev1.PodRunning) || reflect.DeepEqual(aPod.Status.Phase, corev1.PodPending)) && aPod.DeletionTimestamp == nil {
+			if (reflect.DeepEqual(aPod.Status.Phase, corev1.PodRunning) || reflect.DeepEqual(aPod.Status.Phase, corev1.PodPending)) && (aPod.DeletionTimestamp == nil || len(aPod.Finalizers) != 0) {
```
[the remark this answers is not in the document]
Comment (rueian) [reply ray-operator/controllers/ray/suite_helpers_test.go:36]: Without this, the `listResourceFunc` will always return 0 even if Pods are not yet deleted because of their Finalizers, and make the following assertion fail.
```go
Eventually(listResourceFunc(ctx, &allPods, allFilters...), time.Second*3, time.Millisecond*500).
	Should(Equal(numPods), fmt.Sprintf("all pods %v", allPods.Items))
```

I found this failure very misleading and, most importantly, without the change, we can't make sure those Pods still exist.

**Comment by kevin85421** [comment]
What's the YAML you use? I am wondering why "Desired CPU ..." is 0.

**Review by kevin85421 — APPROVED** [review]

**Comment by rueian** [comment]
> What's the YAML you use? I am wondering why "Desired CPU ..." is 0.

I use a RayCluster CR with no k8s resource requests. That's the reason why its CR status shows 0s.

File: ray-operator/controllers/ray/suite_helpers_test.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (rueian) [reply ray-operator/controllers/ray/suite_helpers_test.go:36]: I don't think the `k8sClient.List` does anything special. The behavior you observed was expected but since the deletionTimestamp was not zero, the pod would not be counted to the return value of `listResourceFunc`.

That made `listResourceFunc` always return 0 even though those Pods did still exist. Therefore, I want it to count Pods with non-zero deletionTimestamps and finalizers, otherwise, the 0 returned is too misleading.
